plutus-metatheory/src/Algorithmic/ReductionEC.lagda.md

```diff
@@ -25,40 +27,56 @@ open import Utils hiding (TermCon)
 open import Type using (Ctx⋆;∅;_,⋆_;_⊢⋆_;_∋⋆_;Z)
 open _⊢⋆_
 import Type.RenamingSubstitution as T
-open import Algorithmic using (Ctx;_⊢_;arity;btype;Term;Type;conv⊢)
+open import Algorithmic using (Ctx;_⊢_;conv⊢)
+open import Algorithmic.Signature using (btype;btype-∅;btype-sig2type;_[_]SigTy)
 open Ctx
 open _⊢_
 open import Algorithmic.RenamingSubstitution using (_[_];_[_]⋆)
 open import Algorithmic.Properties using (lem-·⋆;lem-unwrap)
 open import Type.BetaNBE using (nf)
 open import Type.BetaNBE.RenamingSubstitution using (_[_]Nf)
-open import Type.BetaNormal using (_⊢Nf⋆_;embNf;weakenNf)
+open import Type.BetaNormal using (_⊢Nf⋆_;_⊢Ne⋆_;embNf;weakenNf)
 open _⊢Nf⋆_
+open _⊢Ne⋆_
 open import Builtin 
 open import Builtin.Constant.Type Ctx⋆ (_⊢Nf⋆ *) using (TyCon)
 open import Builtin.Constant.Term Ctx⋆ Kind * _⊢Nf⋆_ con using (TermCon)
 open TermCon
+
+open import Builtin.Signature using (Sig;sig;Args;_⊢♯;nat2Ctx⋆;fin2∈⋆;args♯)
+open Sig
+
+open Builtin.Signature.FromSig Ctx⋆ (_⊢Nf⋆ *) nat2Ctx⋆ (λ x → ne (` (fin2∈⋆ x))) con _⇒_ Π 
+    using (sig2type;♯2*;SigTy;sig2SigTy;sigTy2type;saturatedSigTy;convSigTy)
+open SigTy
 ```
 
 ## Values
 
-```
+``` 
 data Value : {A : ∅ ⊢Nf⋆ *} → ∅ ⊢ A → Set
 
-data BApp (b : Builtin) : ∀{az}{as}
-  → az <>> as ∈ arity b
-  → ∀{A} → ∅ ⊢ A → Set where
-  base : ∀{C}(p : C ≡ btype b) → BApp b (start (arity b)) (builtin b / p)
-  step : ∀{A B}{az as}
-    → (p : az <>> (Term ∷ as) ∈ arity b)
-    → {t : ∅ ⊢ A ⇒ B} → BApp b p t
-    → {u : ∅ ⊢ A} → Value u → BApp b (bubble p) (t · u)
-  step⋆ : ∀{C B}{az as}
-    → (p : az <>> (Type ∷ as) ∈ arity b)
-    → {t : ∅ ⊢ Π B} → BApp b p t
-    → {A : ∅ ⊢Nf⋆ K}
+data BApp (b : Builtin) : 
+    ∀{tn tm tt} → {pt : tn ∔ tm ≣ tt}
+  → ∀{an am at} → {pa : an ∔ am ≣ at}
+  → ∀{A} → SigTy pt pa A → ∅ ⊢ A → Set where
+  base : (q : proj₁ (sig2SigTy (signature b)) ≡ btype b) → BApp b (proj₂ (sig2SigTy (signature b))) (builtin b / q )
+  step : ∀{A B}{tn}
+    → {pt : tn ∔ 0 ≣ fv♯ (signature b)} 
+    → ∀{an am}{pa : an ∔ suc am ≣ args♯ (signature b)}
+    → {σB : SigTy pt (bubble pa) B}
+    → {t : ∅ ⊢ A ⇒ B} → BApp b (A B⇒ σB) t
+    → {u : ∅ ⊢ A} → Value u → BApp b σB (t · u)
+  step⋆ : ∀{C B}
+    → ∀{tn tm} {pt : tn ∔ (suc tm) ≣ fv♯ (signature b)} 
+    → ∀{an am}{pa : an ∔ (suc am) ≣ args♯ (signature b)}
+    → {σB : SigTy (bubble pt) pa B}
+    → {t : ∅ ⊢ Π B} → BApp b (sucΠ σB) t
+    → {A : ∅ ⊢Nf⋆ *}
     → (q : C ≡ B [ A ]Nf)
-    → BApp b (bubble p) (t ·⋆ A / q)
+    → {σC : SigTy (bubble pt) pa C}
+    → (σq : σC ≡ substEq (SigTy (bubble pt) pa) (sym q) (σB [ A ]SigTy))
```

Comment:
The problem with this is what `SigTy` to use in the return type?
If we put as `SigTy` the result of the substitution then we have a function on the return type of a datatype (a.k.a.  Green Slime™️). 
So, basically the purpose of the `σq` is the same as the purpose of `q` : to avoid green slime.

Reply:
I see green-slime-avoidance justifies the need for `σC`. I was wondering if you need the proof `σq`? It doesn't seem to be used here and and `σC`, `C` and `B [ A ]Nf` are related by `q` and `σC`. So, I guess my question boils down to - is `σq` used anywhere and if it is can it be derived or do you need to carry it around?

Reply:
If we remove `σq` then `σC` cannot be inferred. Also, I don't know if we will need the proof when I starting doing the behavioural equivalence theorems.